rust/worker/src/execution/operators/merge_metadata_results.rs

```diff
@@ -107,289 +113,144 @@ impl Operator<MergeMetadataResultsOperatorInput, MergeMetadataResultsOperatorOut
+            HashSet::from_iter(mat_records.iter().filter_map(|(log, _)| {
+                matches!(
+                    log.final_operation,
+                    MaterializedLogOperation::DeleteExisting
                 )
-                .await
-                {
-                    Ok(reader) => reader,
-                    Err(e) => {
-                        match *e {
-                            RecordSegmentReaderCreationError::UninitializedSegment => {
-                                // This means no compaction has occured, so we can just return whats on the log.
-                                return Ok(MergeMetadataResultsOperatorOutput {
-                                    ids,
-                                    metadata,
-                                    documents,
-                                });
-                            }
-                            RecordSegmentReaderCreationError::BlockfileOpenError(_) => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                            RecordSegmentReaderCreationError::InvalidNumberOfFiles => {
-                                error!("Error creating Record Segment: {:?}", e);
-                                return Err(
-                                    MergeMetadataResultsOperatorError::RecordSegmentCreationError(
-                                        *e,
-                                    ),
-                                );
-                            }
-                        }
+                .then_some(log.offset_id)
+            }));
+
+        // Merge the offset ids, assuming the user_offset_ids and filtered_offset_ids are ordered.
+        // This should be a sorted list of unique offset ids.
+        // We are treating the sorted vecs as sets.
+        let merged_oids_holder: Vec<u32>;
+        let merged_offset_ids = match (&input.filtered_offset_ids, &input.user_offset_ids) {
+            (Some(fids), Some(uids)) => {
+                // We take the set intersect of offset ids of those passing the filter and those correspond to user specified ids.
+                merged_oids_holder = merge_sorted_vecs_conjunction(fids, uids);
+                &merged_oids_holder
+            }
+            (Some(oids), None) | (None, Some(oids)) => oids,
+            _ => {
+                // If both filter and user offset ids are None, it suggests user did not specify any filter. We should return all available documents.
+                // Offset ids from the materialized records should be unique, so we only need to sort them.
+                let mut live_log_offset_ids = mat_records
```

Comment:
I would prefer if we did not repeatedly iterate over mat_records, this can be quite slow when the log is large. While it would be nice to assume this is not possible, we should be as defensive as we can here.

We already iterate over the log above to generate `deleted_offset_ids` - we can do this in one pass.